plutus-emulator/src/Wallet/Emulator/MultiAgent.hs

```diff
@@ -0,0 +1,220 @@
+    }
+
+-- | Initialise the emulator state with a blockchain.
+emulatorState :: Blockchain -> EmulatorState
+emulatorState bc = emptyEmulatorState
+    & chainState . Chain.chainNewestFirst .~ bc
+    & chainState . Chain.index .~ Index.initialise bc
+
+-- | Initialise the emulator state with a pool of pending transactions.
+emulatorStatePool :: Chain.TxPool -> EmulatorState
+emulatorStatePool tp = emptyEmulatorState
+    & chainState . Chain.txPool .~ tp
+
+-- | Initialise the emulator state with a single pending transaction that
+--   creates the initial distribution of funds to public key addresses.
+emulatorStateInitialDist :: Map PubKey Value -> EmulatorState
+emulatorStateInitialDist mp = emulatorStatePool [tx] where
+    tx = Tx
+            { txInputs = mempty
+            , txOutputs = uncurry (flip pubKeyTxOut) <$> Map.toList mp
+            , txForge = foldMap snd $ Map.toList mp
+            , txFee = mempty
+            , txValidRange = WAPI.defaultSlotRange
+            , txSignatures = mempty
+            , txData = mempty
+            }
+
+type MultiAgentEffs = '[State EmulatorState, Error WAPI.WalletAPIError, Error AssertionError, Chain.ChainEffect]
+
+handleMultiAgent
+    :: forall effs. Members MultiAgentEffs effs
+    => Eff (MultiAgentEffect ': effs) ~> Eff effs
+handleMultiAgent = interpret $ \case
+    -- TODO: catch, log, and rethrow wallet errors?
+    WalletAction wallet act -> act
+        & raiseEnd2
+        & Wallet.handleWallet
+        & subsume
+        & NC.handleNodeClient
+        & interpret (handleZoomedState (walletState wallet))
+        & interpret (handleZoomedWriter p1)
+        & interpret (handleZoomedState (walletClientState wallet))
+        & interpret (handleZoomedWriter p2)
+        & interpret (writeIntoState emulatorLog)
+        where
+            p1 :: Prism' [EmulatorEvent] [Wallet.WalletEvent]
+            p1 = below (walletEvent wallet)
+            p2 :: Prism' [EmulatorEvent] [NC.NodeClientEvent]
+            p2 = below (walletClientEvent wallet)
+    Assertion a -> assert a
+    Failure message -> throwError $ GenericAssertion message
+
+-- | Issue an 'Assertion'.
+assert :: (Members MultiAgentEffs effs) => Assertion -> Eff effs ()
+assert (IsValidated txn)            = isValidated txn
+assert (OwnFundsEqual wallet value) = ownFundsEqual wallet value
+
+-- | Issue an assertion that the funds for a given wallet have the given value.
+ownFundsEqual :: (Members MultiAgentEffs effs) => Wallet.Wallet -> Value -> Eff effs ()
+ownFundsEqual wallet value = do
+    es <- get
+    ws <- case Map.lookup wallet $ _walletStates es of
+        Nothing -> throwError $ GenericAssertion "Wallet not found"
+        Just ws -> pure ws
+    let total = foldMap (txOutValue . txOutTxOut) $ ws ^. Wallet.ownFunds
+    if value == total
+    then pure ()
+    else throwError $ GenericAssertion $ T.unwords ["Funds in wallet", tshow wallet, "were", tshow total, ". Expected:", tshow value]
+    where
```

Comment:
That `if-then-else-where` without any indentation is hard to read to me.

Reply:
This stuff is just moved from the other file, but I agree.